keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,958 @@
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The kubelet does not provide complete information about the container
+resources specification (requests and limits) to CRI. However, various use
+cases have been identified where detailed knowledge of all the resources can be
+utilized in container runtimes for more optimal resource allocation to improve
+application performance and reduce cross-application interference.
+
+This KEP proposes a CRI API extension for disclosing container resource
+requests and limits to container runtimes.
+
+## Motivation
+
+Kubelet manages the native resources (CPU and memory) and communicates
+resource parameters over the CRI API to the runtime. However, the original
+details of the resource spec are lost as they get translated (within kubelet)
+to platform-specific (i.e. Linux or Windows) resource controller parameters
+like cpu shares, memory limits etc. Non-native resources such as extended
```

Comment:
How does this proposal interact with mechanisms such as the existing device plugin API, CDI devices, or DRA in terms of providing access to non-native resources? In which cases is the number of plugin devices requested useful to a CRI runtime?

Maybe my question is, should this proposal be limited to only CPU and Memory, or extended to better capture the acctual resources allocated to all containers? The latter would then include CDI devices from DRA or device plugins, for example.

Reply:
@elezar CPU and Memory is already handled via https://github.com/kubernetes/kubernetes/pull/104886 we need the additional information for non "trivial" devices, like you all already mentioned.

Reply:
In the case of Kata, we need to create the correct PCIe topology during sandbox creation time. We need to know how many PCIe root-ports or PCIe switch-ports to create before the device is hot-plugged in the create container phase. 
In the case of cold-plugging the PCIe topology and direct attach of the devices needs to happen at sandbox creation time not container.

Reply:
Thanks for the clarification. I suppose my question is whether just listing the devices as requested is sufficient to extract the information that you need. For example, for a node with multiple devices I don't see how the the count for `vendor.com/xpu` in the request maps to PCI topology -- unless the Container Runtime is also aware of device specifics.

Note, I'm not opposed to passing down the information as requested if this really is useful, I would just like the proposal extended to properly capture the resource request details -- ideally reusing the existing structures to ensure these remain consistent.

Reply:
Ideally, each extended resource allocation or resource claim is a CDI device, and the name of the allocated resource is a fully qualified CDI name rather than some "logical" name like `vendor.com/xpu: 1`. 
With this information, a runtime like Kata can get the correct information from the CDI registry since the primary key is the FQN. 
If the allocated resource is "just"  a logical name, we need to provide additional information because we cannot deduce the correct information from an extended resource name, and we do not know anything about the DevicePlugin business logic.

Reply:
There is also another complication which depends on how we're presenting the device into a VM. (1) The VFIO device is handled by the native driver or (2) the VFIO device is handled by the `vfio-pci` driver in the VM itself. 
For the second case, we're just passing-through the VFIO IOMMU group and a new one is created in the VM.
For the first case when a native driver handles the HW, how do we know which instance was meant to be allocated by the container if multiple devices are allocated?

Reply:
@elezar do you think this has been addressed (and marked as resolved?). To recap, the KEP now contains all mounts, devices and CDI devices.

Reply:
Yes, I think the new itteration looks better in this regard. The key being that the resources listed includes those under the control of the various device plugins for extended resources.

Reply:
Resolved?